### PR TITLE
TDC simulation for HFQIE10 (12_1)

### PR DESCRIPTION
**Pull request by mseidel42**
#### PR description:

* This PR adds TDC timing simulation for the HF --> enables the use of data noise filters in MC
  * Replaces 11_3 PR https://github.com/cms-sw/cmssw/pull/33829
  * Now precise samples are stored in PreMixStage1 and converted back to usual DIGIs in PreMixStage2
* Original presentation in HCAL DPG: https://indico.cern.ch/event/1026044/contributions/4324030/attachments/2227968/3774461/qie10tdc_mseidel.pdf
    * Pulse shape deemed sufficient (no update)
    * Eta-dependence fixed by new HF shower library

#### PR validation:

Tested with workflows 11634.0 (Run3 ttbar), 11834.0 (Run3 ttbar+PU), 250202.181 (Run2 ttbar+premix PU), no "Mismatched calo signals" messages.

MC Run3 ttbar+PU:
image

MC Run2 ttbar+premix PU (time alignment for Run2 MC is a little off, keep frozen):
image

Run2 data:
image

@abdoulline @kpedro88 @tvami

### CODE REVIEW

**Comment by kpedro88** [comment]
please test

**Comment by kpedro88** [comment]
Are the MC hits with TDC=31 expected? These don't seem to occur in data.

**Comment by mseidel42** [comment]
> Are the MC hits with TDC=31 expected? These don't seem to occur in data.

Those are the special codes, just divided by 2 to compare with the data plot (ns instead of TDC).
https://github.com/cms-sw/cmssw/blob/a87b522f54f81316827d5fe365b17ba36fb19a63/SimCalorimetry/HcalSimAlgos/interface/HcalTDCParameters.h#L12-L15

**Comment by tvami** [comment]
@cmsbuild , please abort

**Comment by tvami** [comment]
test parameters:
  - workflows = 11634.0,11834.0,250202.181

**Comment by tvami** [comment]
@cmsbuild , please test

**Comment by tvami** [comment]
> @tvami for future reference, workflows in the short matrix are always tested, so it's not necessary to specify them explicitly.

@kpedro88 I had a look at another PR to see what's included
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-2d40d9/18638/runTheMatrix-results/
11634.0 and 250202.181 is indeed there, but it seems 11834.0 is not part of the short matrix

**Comment by kpedro88** [comment]
Easier to check https://github.com/cms-sw/cmssw/blob/04e25fafc456b764f3b59c4e88b1c0f1b681e368/Configuration/PyReleaseValidation/scripts/runTheMatrix.py#L61-L100

Indeed, maybe a Run3 PU workflow should be considered for regular testing.

**Comment by abdoulline** [comment]
@mseidel42   
Once now MC delivers TDC (which are quite different from data ones, so that thier timing could be less than ~5ns),  there might be undesirable effect on MC HF reconstruction, unless I miss (forgot)  something? 
Adding  @toropin  @igv4321

**Comment by abdoulline** [comment]
@toropin @igv4321 
Please, remind us - what is there for MC as HFPhase1PMTParams content? 
https://twiki.cern.ch/twiki/bin/view/CMS/HFPhase1PMTParams 
- Should be something like "all pass through"?

**Comment by tvami** [comment]
Are the changes supposed to effect Phase-2 as well?
https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_1_X_2021-09-16-1100+29891e/45509/34834.999_TTbar_14TeV+2026D76PU_PMXS1S2PR+TTbar_14TeV_TuneCP5_GenSimHLBeamSpot14+PREMIX_PremixHLBeamSpot14PU+DigiTriggerPU+RecoGlobalPU+HARVESTGlobalPU/

**Comment by abdoulline** [comment]
@tvami 
If HF QIE10 (since 2017) digitization is somehow affected,  it's for >= 2017  (2018, Run3 and Phase2).

I have the impression that the changes are very small, of the order of "numerical precision" (conversion of preciseSamples to the regular ones?), so that there is effectively no filtering (TDC discrimination)  in HF Reco for MC.

**Comment by tvami** [comment]
I'm not really sure what to look at...
Looking into JetMet: https://tinyurl.com/ydv5enj9
b-tagging: https://tinyurl.com/yz9gkl9l
HCalDigisV: https://tinyurl.com/yz3b8lvm
everything just looks like statistical fluctuations, but it's hard to tell with these limited statistic

**Comment by tvami** [comment]
Should maybe the era modifier be changed so that all this is not happening for the Phase-2 samples?
https://github.com/cms-sw/cmssw/blob/a87b522f54f81316827d5fe365b17ba36fb19a63/SimCalorimetry/HcalSimProducers/python/hcalSimParameters_cfi.py#L213-L218

**Comment by kpedro88** [comment]
@tvami since the HF is part of the Phase 2 detector, I'm not sure why it would make sense to exclude updates to the HF simulation from Phase 2 workflows.

**Comment by tvami** [comment]
> @tvami since the HF is part of the Phase 2 detector, I'm not sure why it would make sense to exclude updates to the HF simulation from Phase 2 workflows.

Ok, but is there any real change expected from this PR for Phase-2? To my understanding, no. But there are some changes, which are claimed to be statistical fluctuations, however the very limited statistical bin-by-bin comparisons are not conclusive. So either there should be some higher stat studies done for Phase-2 showing that this PR's changes are not relevant, or the Phase-2 people give their blessings without further studies and I just trust that :)

**Comment by abdoulline** [comment]
Single pion ("CaloScan") 2021 test w/wo PR shows tiny difference for HF (4-th digit in the mean values of the HF distribitions):
 
https://cms-cpt-software.web.cern.ch/cms-cpt-software/General/Validation/SVSuite/HCAL/calo_scan_single_pi/12_1_X/12_1_X_2021-09-16-2300_PR35308_run3_vs_12_1_X_2021-09-16-2300_run3_SinglePi/

(1)  HF RecHits spectrum 
RecHits_energy_HF


(2) The good thing is the HF reco timing, which now looks quite reasonable 
RecHitsTask_timing_vs_energy_profile_HF



Once anything changed upstream, then Jet iteration (sequence) may change.

**Comment by toropin** [comment]
I did not participate in the preparation of the conditions for the MC.
I would like to remind you that we are discussing two cuts here.
One by the time of arrival of the signal: TDC cut, and the second by the difference
of signals from the two anodes.
If you do not make these cuts transparent to the MC, then they can have some effect on the MC.
As a reminder, I am sending a couple of slides from old report.

[X][X]
[X]

**Comment by abdoulline** [comment]
@toropin  could you, please, edit your post here (directly at https://github.com/cms-sw/cmssw/pull/35308)
There is no attachment and an inclusion of my message looks quite "bulky"...
Here we discuss HF TDC aspect (which has been changed by this PR) specifically for MC. 
It looks like MC tag was/is "pass through", as there seems to be no difference w/wo this PR in the number of HF RecHits...

**Comment by mseidel42** [comment]
Hi all, thanks a lot for the checks! This should be relevant as long as HF lives, so it definitely needs to go into Phase-2 as well.

I was wondering why there are small differences in noPU WFs.
See for example this DIGI printout, with PR:
```
DetId 1228971067, 3 samples, 150 preciseSamples, 50 preciseStep
0:17.0754 precise: 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
1:30.2014 precise: 0 0 0 0 0 0 0 0 0.00720502 0.00720502 0.00720502 0.00720502 0.0570808 0.0570808 0.262303 0.262303 0.69916 0.69916 1.08096 1.08096 1.08081 1.08081 0.957925 0.957925 0.769182 0.769182 0.597649 0.597649 0.464369 0.464369 0.360811 0.360811 0.280348 0.280348 0.217828 0.217828 0.169251 0.169251 0.131507 0.131507 0.10218 0.10218 0.0793929 0.0793929 0.0616877 0.0616877 0.0479309 0.0479309 0.0372419 0.0372419
2:13.1251 precise: 0.0289367 0.0289367 0.0224836 0.0224836 0.0174696 0.0174696 0.0135738 0.0135738 0.0105467 0.0105467 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
```
and without PR:
```
DetId 1228971067, 3 samples
0:17.0754 precise:
1:30.187 precise:
2:13.1251 precise:
```
--> 30.2014 vs 30.187 fC.

Digging further in the code, I think these tiny numerical differences come from the fact that the precise samples bypass the `CaloCachedShapeIntegrator` which internally uses rounding to full ns.

**Comment by abdoulline** [comment]
@mseidel42  thanks for tracking down the numerical difference !
(which I previously/above assumed as somehow coming from integration of preciseSamples).  
It's certainly way smaller than any other uncertainties.

**Comment by igv4321** [comment]
@abdoulline The MC HFPhase1PMTParams table is "all pass". The essential piece of code used to construct it is
https://cmssdt.cern.ch/lxr/source/CondTools/Hcal/test/make_HFPhase1PMTParams_dummy.cc

**Comment by abdoulline** [comment]
> @abdoulline The MC HFPhase1PMTParams table is "all pass". The essential piece of code used to construct it is
> https://cmssdt.cern.ch/lxr/source/CondTools/Hcal/test/make_HFPhase1PMTParams_dummy.cc

@igv4321 
OK, thank you for confirmation!

**Comment by civanch** [comment]
+1

it seems that the discussion is converged.

**Comment by tvami** [comment]
+alca 
 * changes for the Phase-1 detector as expected
 * changes for Phase-2 are understood as well, Phase-2 team has been notified

File: SimCalorimetry/CaloSimAlgos/interface/CaloHitResponse.h [hunk 1]
```diff
@@ -134,6 +144,8 @@ class CaloHitResponse {
 
   double thePhaseShift_;
   bool storePrecise;
+  bool PreMixDigis;
```
Comment (perrotta) [inline SimCalorimetry/CaloSimAlgos/interface/CaloHitResponse.h:147]: Please use lowercase initials for class members

File: SimCalorimetry/CaloSimAlgos/interface/CaloHitResponse.h [hunk 2]
```diff
@@ -134,6 +144,8 @@ class CaloHitResponse {
...
+  bool HighFidelityPreMix;
```
Comment (perrotta) [inline SimCalorimetry/CaloSimAlgos/interface/CaloHitResponse.h:148]: same here

File: SimCalorimetry/CaloSimAlgos/interface/CaloHitResponse.h [hunk 3]
```diff
@@ -35,9 +35,17 @@ class CaloHitResponse {
   typedef std::map<DetId, CaloSamples> AnalogSignalMap;
   // get this from somewhere external
   enum { BUNCHSPACE = 25 };
-
-  CaloHitResponse(const CaloVSimParameterMap *parameterMap, const CaloVShape *shape);
-  CaloHitResponse(const CaloVSimParameterMap *parameterMap, const CaloShapes *shapes);
+  const double dt = 0.5;
```
Comment (perrotta) [inline SimCalorimetry/CaloSimAlgos/interface/CaloHitResponse.h:38]:
```suggestion
  constexpr double dt = 0.5;
  constexpr int invdt = 2;
```

Comment (mseidel42) [reply SimCalorimetry/CaloSimAlgos/interface/CaloHitResponse.h:38]: This leads to `error: non-static data member 'dt' declared 'constexpr'`

Shall I use `static constexpr` instead?

Comment (abdoulline) [reply SimCalorimetry/CaloSimAlgos/interface/CaloHitResponse.h:38]: I'm wondering why not just 
static const (int/double) ?

Comment (perrotta) [reply SimCalorimetry/CaloSimAlgos/interface/CaloHitResponse.h:38]: I would use` static constexpr `if so...
But honestly I did not try it: just do whatever does not give errors

Comment (mseidel42) [reply SimCalorimetry/CaloSimAlgos/interface/CaloHitResponse.h]: Ok, that seems to work fine

**Comment by tvami** [comment]
@cmsbuild , please test

**Comment by tvami** [comment]
+alca

**Comment by civanch** [comment]
+1

**Comment by perrotta** [comment]
+1

**Pull request merged**